pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -103,6 +106,17 @@ func (d *PmemCSIDriver) Reconcile(r *ReconcileDeployment) (bool, error) {
 		// Deployment successfull, so no more reconcile needed for this deployment
 		return false, nil
 	case api.DeploymentPhaseRunning:
+		if !foundInCache {
+			// Possible that operator restarted, so we can't assume that
+			// everything upto date. Look at running deployment if any changes
+			// needed
+			oldDeployment, err := d.buildFromPreDeployed(r)
+			if err != nil {
+				return true, err
+			}
+
+			changes = d.Compare(oldDeployment)
```

Comment:
I suspect there is one situation that isn't handled by just looking at the deployment spec.

Consider the case where we upgrade the operator and that the new operator will change the driver deployment somehow, for example by adding a new `--foobar` parameter for one of the sidecars. This difference between the running driver deployment and the one that the new operator would create is not detected, is it?

What should happen is that the operator should update the objects that it creates such that they match exactly what it would create from scratch. The simplest (?) solution would be to just do that unconditionally each time the operator starts. The apps operator can then check whether it really needs to restart pods.

This logic hinges on both operator versions creating exactly the same objects. At one point me might add some new object (for example, a service) that the old operator doesn't know about. In case of a downgrade, it then would leave that service object unchanged although it's not supposed to be part of the downgraded driver deployment.

I don't have a good solution for this because there is no generic "list all objects with certain labels" API in Kubernetes. One has to do that for all relevant object types, and "relevant" is unknown for a downgrade (the older operator would need to know what will be added in the future). However, we can implement this for an upgrade because then we know all types that may have been created in previous versions.

Reply:
> I suspect there is one situation that isn't handled by just looking at the deployment spec.
> Consider the case where we upgrade the operator and that the new operator will change the driver deployment somehow, for example by adding a new --foobar parameter for one of the sidecars. This difference between the running driver deployment and the one that the new operator would create is not detected, is it?

Wouldn't be better to separate normal operator restart from version upgrade? say for version upgrades we could deal by adding an annotation to Deployment object, so on recocile we compare versions and decide if needs to re-deploy a Deployment.

Reply:
On re-deployment we might loose find and reject if any conflicting deployment changes. Say, driverMode/pmemPercentage changed when the operator is not running.

Reply:
So for invariant attributes we must still check the running driver deployment and update the deployment object accordingly. But then we can simply update all generated objects.

Reply:
Made the changes as suggested, so now deployment reconcile on operator restart refreshes the objects and checks if any incompatible changes to be rejected.